File: src/main/java/macro/library/database/BookTable.java
package macro.library.database;

import macro.library.Isbn;
import macro.library.book.Book;
import macro.library.book.Format;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.sql.ResultSet;
import java.sql.SQLException;

/**
 * Created by Macro303 on 2019-Oct-22
 */
public class BookTable extends IdTable<Book, Isbn> {
	@NotNull
	private static final Logger LOGGER = LogManager.getLogger(BookTable.class);
	@NotNull
	public static final BookTable INSTANCE = new BookTable();

	private BookTable() {
		super("book", "isbn");
	}

	@Override
	public boolean insert(@NotNull Book item) {
		var query = String.format("INSERT INTO %s(isbn, title, subtitle, author, publisher, format) VALUES(?, ?, ?, ?, ?, ?);", tableName);
		return insert(query, item.getISBN(), item.getTitle(), item.getSubtitle(), item.getAuthor(), item.getPublisher(), item.getFormat().ordinal());
	}

	@Override
	public boolean update(@NotNull Book item) {
		var query = String.format("UPDATE %s SET title = ?, subtitle = ?, author = ?, publisher = ?, format = ? WHERE isbn = ?;", tableName);
		return update(query, item.getTitle(), item.getSubtitle(), item.getAuthor(), item.getPublisher(), item.getFormat().ordinal(), item.getISBN());
	}

	@Override
	protected void createTable() {
		var query = String.format("CREATE TABLE %s(isbn TEXT PRIMARY KEY NOT NULL UNIQUE, title TEXT, subtitle TEXT, author TEXT NOT NULL, publisher TEXT NOT NULL, format INTEGER NOT NULL DEFAULT(0));", tableName);
		insert(query);
	}

	@Nullable
	@Override
	protected Book parse(@NotNull ResultSet result) throws SQLException {
		return new Book(Isbn.of(result.getString("isbn")), result.getString("title"), result.getString("subtitle"), result.getString("author"), result.getString("publisher"), Format.values()[result.getInt("format")]);
	}
}

File: src/main/kotlin/macro/library/database/BookTable.kt
package macro.library.database

import macro.library.book.Book
import macro.library.book.Format
import macro.library.book.Isbn
import org.apache.logging.log4j.LogManager
import java.sql.ResultSet
import java.sql.SQLException
import java.util.*

/**
 * Created by Macro303 on 2019-Oct-31
 */
object BookTable : IdTable<Book, Isbn>(tableName = "book", idName = "isbn") {
	private val LOGGER = LogManager.getLogger(BookTable::class.java)

	override fun insert(item: Book): Boolean {
		val query = "INSERT INTO $tableName(isbn, title, subtitle, publisher, format, image_small, image_medium, image_large) VALUES(?, ?, ?, ?, ?, ?, ?, ?);"
		return insert(item.isbn, item.title, item.subtitle, item.publisher, item.format.ordinal, item.imageSmall, item.imageMedium, item.imageLarge, query = query)
	}

	override fun update(item: Book): Boolean {
		val query = "UPDATE $tableName SET title = ?, subtitle = ?, publisher = ?, format = ?, image_small = ?, image_medium = ?, image_large = ? WHERE isbn = ?;"
		return update(item.title, item.subtitle, item.publisher, item.format.ordinal, item.imageSmall, item.imageMedium, item.imageLarge, item.isbn, query = query)
	}

	override fun createTable() {
		val query = "CREATE TABLE $tableName(" +
				"isbn TEXT PRIMARY KEY NOT NULL UNIQUE, " +
				"title TEXT NOT NULL, " +
				"subtitle TEXT, " +
				"publisher TEXT NOT NULL, " +
				"format INTEGER NOT NULL DEFAULT(0)," +
				"image_small TEXT," +
				"image_medium TEXT," +
				"image_large TEXT);"
		insert(query = query)
	}

	@Throws(SQLException::class)
	override fun parse(result: ResultSet): Book {
		return Book(
			Isbn.of(result.getString("isbn"))!!,
			result.getString("title"),
			result.getString("subtitle"),
			result.getString("publisher"),
			Format.values()[result.getInt("format")],
			result.getString("image_small"),
			result.getString("image_medium"),
			result.getString("image_large")
		)
	}

	fun search(title: String? = null, subtitle: String? = null, author: String? = null): List<Book> {
		var query = "SELECT * FROM $tableName"
		val params = ArrayList<String>()
		if (title != null) {
			query += " ${if (query.endsWith(tableName)) "WHERE" else "AND"} title LIKE ?"
			params.add(title)
		}
		if (subtitle != null) {
			query += " ${if (query.endsWith(tableName)) "WHERE" else "AND"} subtitle LIKE ?"
			params.add(subtitle)
		}
		if (params.isNotEmpty())
			query += " COLLATE NOCASE"
		return search(*params.toTypedArray(), query = "$query;")
	}
}